### ISSUE TITLE
Reading stdout and stderr while a process created with Bun Shell is running

### ISSUE THREAD

**Issue by jamesbradlee** [oven-sh/bun#14693]
### What is the problem this feature would solve?

Currently, it is not possible to read from `stdout` or `stderr` while a process created with Bun Shell is running. The current implementation of Bun Shell as of writing, when calling `lines()`, `text()`, etc. methods use a form `await this.quiet()`, essentially waiting for the process to finish before reading the buffered `stdout` and `stderr` streams.

### What is the feature you are proposing to solve the problem?

Add `stdout` and `stderr` streams to the `ShellPromise` class.

```ts
class ShellPromise {
  stdout: ReadableStream<Uint8Array | Buffer | string | Blob | etc>
  stderr: ReadableStream<Uint8Array | Buffer | string | Blob | etc>
}
```

> [!NOTE]
>
> I do not mean that the `stdout` and `stderr` readable streams should output all of them, but that a either of them should be picked.

### What alternatives have you considered?

In the meantime I have implemented my own solution that works by accepting `(template: TemplateStringsArray, ...args: string[])`. Turn it back into a string, spawning a process using `node:child_process`' `spawn("/bin/sh", ["-c", commandStr], { ...options })`. It's not as safe as Bun Shell, doesn't have as many features, but it supports stdout and stderr streams.

**Comment by testersen** [comment]
I realize this may be difficult to implement, because in its current form, the process is not actually started before `then()` is called. So the `stdout` and `stderr` properties must be created before `.then()` and when calling `.then()` the process will start piping values to the streams. That way the properties exists before calling `.then()`.

**Comment by guest271314** [comment]
I think this is already possible by piping the output of the command to the PID of the current Bun process, then reading `Bun.file("/dev/stdin").stream()`.

**Comment by guest271314** [comment]
Here's a couple different ways to do this.

The first approach writes to a flat file.

The second approach uses `node:net` with a Unix domain socket.

`shell.js` launches the Bun shell script. `stream.js` is the Bun shell shell, and optionally also uses the Bun shell.

#### Communication using a flat file

`shell.js 30`

```
#!/usr/bin/env -S /home/user/bin/bun run
import { $ } from "bun";
import process from "node:process";
import { unlink, watch } from "node:fs/promises";

const path = "./bun-shell.sock";
const watcher = watch(import.meta.dir);

let file = Bun.file(path);

if (await file.exists()) {
  await unlink(path);
}

(async () => {
  let change = false;
  for await (const event of watcher) {
    if (await Bun.file(path).exists()) {
      console.log(await Bun.file(path).text());
    }
  }
})();

await $`./stream.js ${Bun.argv.pop()}`.quiet();

await unlink(path);
process.exit(0);
```

`stream.js`

```
#!/usr/bin/env -S /home/user/bin/bun run
import { $ } from "bun";
import process from "node:process";
import { unlink } from "node:fs/promises";

const path = "./bun-shell.sock";

for (let i = 0; i < 100; i++) {
  await new Promise((resolve) => setTimeout(resolve, Number(Bun.argv.at(-1))));
  //await $`echo ${i} > ${file}`.quiet(); //  >> stdout
  let file = Bun.file(path);
  if (await file.exists()) {
    await unlink(path);
  }
  file = Bun.file(path);
  await Bun.write(file, i);
  // Too slow.
  // await $`echo ${i} > ${file}`.quiet();
}

process.exit(0);

```

#### Communication using `node:net`

`shell.js 30`
```
#!/usr/bin/env -S /home/user/bin/bun run
import { $ } from "bun";
import process from "node:process";
import { createServer } from "node:net";
import { unlink } from "node:fs/promises";

const path = "./bun-shell.sock";

const server = createServer((stream) => {
  stream.on("data", async (data) => {
    Bun.write(Bun.stdout, `${data} `);
  });
  stream.on("end", async () => {
    await unlink(path);
    await Bun.write(Bun.stdout, "Stream closed.");
    server.close();
  });
});

server.listen(path);

await $`./stream.js ${Bun.argv.pop()}`.quiet();

process.exit(0);
```

`stream.js`

```
#!/usr/bin/env -S /home/xubuntu/bin/bun run
import { $ } from "bun";
import process from "node:process";
import { connect } from "node:net";

const stream = connect("./bun-shell.sock");

for (let i = 0; i < 100; i++) {
  await new Promise((resolve) => setTimeout(resolve, Number(Bun.argv.at(-1))));
  stream.write(`${i}`);
}

stream.end();

process.exit(0);
```

**Comment by testersen** [comment]
Thanks for the suggestions, @guest271314, but honestly, this workaround feels like an overengineered solution to a problem that should be simple. Using flat files or Unix domain sockets just to read stdout and stderr while a process is running turns something basic into a complex mess.

According to Bun’s own documentation, "Bun Shell makes shell scripting with JavaScript & TypeScript fun." I think any developer would agree that simplicity is key to making that happen. Simplicity enhances the developer experience, which in turn improves the overall user experience.

It’s not that I couldn’t implement this — I’ve already built an alternative to Bun Shell that supports what I need. It’s built upon `node:child_process` by spawning a `/bin/bash` process and then converting `stdout` and `stderr` into Web Standard `ReadableStreams`, rather than using Node's `ReadableStream`.

But no one should have to resort to this kind of workaround for something so fundamental. A native implementation would keep things clean, efficient, and true to Bun Shell's

**Comment by Jarred-Sumner** [comment]
@testersen yes you are right, Bun Shell should make it easy to stream stdout/stderr without waiting for the shell invocation to exit first

**Comment by testersen** [comment]
Understandable, @Jarred-Sumner! Making a runtime doesn't sound like easy work - especially also not the tons of tools surrounding it!

**Comment by guest271314** [comment]
> Thanks for the suggestions, @guest271314, but honestly, this workaround feels like an overengineered solution to a problem that should be simple. 

Engineered just enough to work.

> It’s built upon `node:child_process` by spawning a `/bin/bash` process and then converting `stdout` and `stderr` into Web Standard `ReadableStreams`, rather than using Node's `ReadableStream`.

That's a reasonable implementation. 

I do that, too, for testing Native Messaging hosts. 

The Bun shell is basically a subprocess, anyway.

**Comment by guest271314** [comment]
> The only reason this hasn't been done yet is we have a lot of stuff on our plate right now and need to hire more people

Make me an offer I can't refuse. 

You should know my work by now. 

I'm gonna test your gear until it breaks, and create at least two (2) different implementations for whatever concept I have, or somebody else has, or a workaround for something that is already broken/not implemented - without getting emotionally invested in your idea being correct. Just the code.

**Comment by guest271314** [comment]
@testersen FWIW this is what I use to test Node.js, Bun, C, C++, Rust, WASM, Amazon Web Services Labs `llrt`, Google `d8`, `SpiderMonkey `js`, Bash Native Messaging hosts, see [How to test the different hosts #2](https://github.com/guest271314/NativeMessagingHosts/discussions/2).

I have not explored trying to do communication with [`dax.js`](https://github.com/dsherret/dax) or Google's `zx` cf. Bun's shell.

```
// deno run -A nm_standalone_test.js ./nm_nodejs.js native-messaging-extension://xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx/

const [path, allowed_origin] = Deno.args;

const command = new Deno.Command(path, {
  args: [allowed_origin],
  stdout: "piped",
  stdin: "piped",
});

const subprocess = command.spawn();
const buffer = new ArrayBuffer(0, { maxByteLength: 1024 ** 2 });
const view = new DataView(buffer);
const encoder = new TextEncoder();
const decoder = new TextDecoder();

function encodeMessage(message) {
  return encoder.encode(JSON.stringify(message));
}

async function* getMessage(readable) {
  let messageLength = 0;
  let readOffset = 0;
  for await (let message of readable) {
    if (buffer.byteLength === 0 && messageLength === 0) {
      buffer.resize(4);
      for (let i = 0; i < 4; i++) {
        view.setUint8(i, message[i]);
      }
      messageLength = view.getUint32(0, true);
      console.log({ messageLength });
      message = message.subarray(4);
      buffer.resize(0);
    }
    if (message.length) {
      buffer.resize(buffer.byteLength + message.length);
      for (let i = 0; i < message.length; i++, readOffset++) {
        view.setUint8(readOffset, message[i]);
      }
      if (buffer.byteLength === messageLength) {
        yield new Uint8Array(buffer);
        messageLength = 0;
        readOffset = 0;
        buffer.resize(0);
      }
    }
  }
}

(async () => {
  try {
    for await (const message of getMessage(subprocess.stdout)) {
      console.log(JSON.parse(decoder.decode(message)));
    }
  } catch (e) {
    console.log(e.message);
    Deno.exit();
  }
})();

let data = encodeMessage(Array(209715)); 

await new Blob([
  new Uint8Array(new Uint32Array([data.length]).buffer),
  data,
])
  .stream()
  .pipeTo(subprocess.stdin, { preventClose: true });

await new Promise((resolve) => setTimeout(resolve, 20));

data = encodeMessage("test");
await new Blob([
  new Uint8Array(new Uint32Array([data.length]).buffer),
  data,
])
  .stream()
  .pipeTo(subprocess.stdin, { preventClose: true });

await new Promise((resolve) => setTimeout(resolve, 20));

data = encodeMessage("");
await new Blob([
  new Uint8Array(new Uint32Array([data.length]).buffer),
  data,
])
  .stream()
  .pipeTo(subprocess.stdin, { preventClose: true });

await new Promise((resolve) => setTimeout(resolve, 20));

data = encodeMessage(1);
await new Blob([
  new Uint8Array(new Uint32Array([data.length]).buffer),
  data,
])
  .stream()
  .pipeTo(subprocess.stdin, { preventClose: true });

await new Promise((resolve) => setTimeout(resolve, 20));

data = encodeMessage(new Uint8Array([97]));
await new Blob([
  new Uint8Array(new Uint32Array([data.length]).buffer),
  data,
])
  .stream()
  .pipeTo(subprocess.stdin, { preventClose: true});

await new Promise((resolve) => setTimeout(resolve, 20));

subprocess.kill("SIGTERM");
```

**Comment by testersen** [comment]
> Engineered just enough to work.

Over engineered :smile:

> You should know my work by now.

By that you mean; changing a comment in a code file, updating a variable name, adding a word to a markdown file, updated a link in a markdown file and added a code example for import attributes in a markdown file?

Getting paid for open-source work might be difficult, I believe and have seen, that paid jobs for open-source usually goes to the people with most valuable contributions - as a "thank you" for the passion in the open source project. Last time I saw this was on Deno, where the top contributors eventually was hired by Deno when PE-firms invested in it.

I didn't mean to be rude, I realize it might come across that way - just wanted to help manage expectations 😅 Good luck anyway 😄

**Comment by guest271314** [comment]
I think Bun has an IPC channel built in to the subprocess creation. I don't think we necessarily get anything special using Bun's shell for a subprocess. 

If you check `process.pid` and `process.ppid` from the script launched within the shell, you'll see that the script PPID is the same as the original script that launched the Bun shell process.

**Comment by guest271314** [comment]
> By that you mean; changing a comment in a code file, updating a variable name, adding a word to a markdown file, updated a link in a markdown file and added a code example for import attributes in a markdown file?

No. I mean testing your code until it breaks and creating solutions to concepts and fixing broken code. Basically all of my GitHub repositories fix bugs marked `WontFix`.

But hey, if that ain't good enough I'm doing great contributing to build multi-million dollar homes (e.g., a recent one for one of the original Microsoft investors) right now. I hack for sport.

**Comment by MatthewAlner** [comment]
➕1 - This would be a neat addition, i'm running some ffmpeg commands and it would be nice if I could parse out the progress.

**Comment by robobun** [comment]
Same feature request as #8365: Bun Shell ($) only hands stdout to JS after the whole shell invocation exits (lines()/text() await the buffered result), and both issues ask for the output to be readable incrementally while the command is still running; exposing stdout/stderr as streams here and a streaming lines() there are two API shapes for the same capability, and one implementation would resolve both.

Duplicate of #8365

Please add a 👍 or any extra details over there so it all stays in one place. If this is actually a different problem, say so here and we'll reopen.
